### PR TITLE
feat (Beta ): improve lock/unlock edit breakout

### PR DESCRIPTION
**Pull request by JudyZhuHz**
# COMPLETES #< INSERT LINK TO ISSUE >

https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-405578

< DESCRIBE THE CONTEXT OF THE ISSUE >

lock/unlock edit breakout

< DESCRIBE YOUR CHANGES >

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [X] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [] I discussed changes with code owners prior to submitting this pull request

- [ ] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 1]
```diff
@@ -34,9 +34,11 @@ const Breakouts = WebexPlugin.extend({
     status: 'string', // only present when in a breakout session
     url: 'string', // appears from the moment you enable breakouts
     locusUrl: 'string', // the current locus url
-    breakoutServiceUrl: 'string', // the current breakout resouce url
+    breakoutServiceUrl: 'string', // the current breakout resource url
     mainLocusUrl: 'string', // the locus url of the main session
     groups: 'array', // appears when create breakouts
+    editLock: 'object', // appears when getBreakout info editlock = true
+    intervalID: 0,
```
Comment (Coread) [inline packages/@webex/plugin-meetings/src/breakouts/index.ts:41]: the type of this should be `"number"` not 0

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts [hunk 2]
```diff
@@ -703,5 +703,117 @@ describe('plugin-meetings', () => {
...
+        breakouts.unLockEditBreakout();
+        assert.calledOnceWithExactly(webex.request, {
+          method: 'DELETE',
+          uri: 'url/editlock/2ad57140-01b5-4bd0-a5a7-4dccdc06904c'
+        });
+
+      });
+    });
+
+    describe('keepEditLockAlive', () => {
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts:795]: issue: there should be a test for making sure the edit lock is cleared if the request fails

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts [hunk 3]
```diff
@@ -703,5 +703,117 @@ describe('plugin-meetings', () => {
...
+
+      it('keep edit lock', () => {
+
+        const clock = sinon.useFakeTimers()
+
+        breakouts.set('editLock', {
+          ttl: 30,
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts:802]: issue: also need a test when ttl is less than 30, to check that the minimum of 30 is used as well as a test when it is more than 30 e.g. 50 to check that 50 is used

**Review by Coread — CHANGES_REQUESTED** [review]
Please see comments

File: packages/@webex/plugin-meetings/src/breakouts/index.ts [hunk 4]
```diff
@@ -503,9 +534,96 @@ const Breakouts = WebexPlugin.extend({
...
+    }
+  },
+
+  /**
+   * keep edit lock alive
+   * @returns {void}
+   */
+  keepEditLockAlive() {
+    if (this.editLock && !!this.editLock.token) {
+      const ttl = this.editLock.ttl < 30 ? 30 : this.editLock.ttl;
```
[the remark this answers is not in the document]
Comment (JudyZhuHz) [reply packages/@webex/plugin-meetings/src/breakouts/index.ts:585]: OK, i will update it

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts [hunk 5]
```diff
@@ -737,6 +737,185 @@ describe('plugin-meetings', () => {
...
+        breakouts.set('editLock', {
+          ttl: 50,
+          token: 'token',
+          state: 'UNLOCKED',
+        });
+
+        breakouts.keepEditLockAlive();
+        clock.tick(25001);
+
+        assert.calledOnceWithExactly(webex.request, {
```
[the remark this answers is not in the document]
Comment (JudyZhuHz) [reply packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts]: Got it, thanks for the reminder. These cases have been added.

File: packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts [hunk 6]
```diff
@@ -737,6 +737,207 @@ describe('plugin-meetings', () => {
...
+          token: 'token',
+          state: 'UNLOCKED',
+        });
+
+        breakouts.keepEditLockAlive();
+        clock.tick(14999);
+
+        assert.notCalled(webex.request);
+
+        clock.tick(15000);
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/breakouts/index.ts:930]: clock.tick moves the timer on by that amount. You don't need to clock tick 1500 here, simply ticking by 1 should have worked

**Review by Coread — APPROVED** [review]
